Fix/design system component

Mohamed-Ashraf-2210 opened the pull request:
refactoring and fixing design system components to enhance consistency, flexibility, and user interaction.

<img width="654" height="339" alt="MainInfoCard" src="https://github.com/user-attachments/assets/565ac271-b470-4c67-ba4f-44c4a4c3c626" />
<img width="383" height="326" alt="GallerySection" src="https://github.com/user-attachments/assets/86c78cc7-00d4-492c-98e5-81323f52fafc" />
<img width="337" height="215" alt="MainDetails" src="https://github.com/user-attachments/assets/994c3dc9-4fe5-4a29-a00a-9429c7e9de20" />
<img width="439" height="226" alt="ReviewCard" src="https://github.com/user-attachments/assets/6074910a-4f4e-4daf-8156-45dc5ef9c42c" />
<img width="985" height="370" alt="SeasonCard" src="https://github.com/user-attachments/assets/598b3eb7-93d8-4502-ba26-011c45be217d" />
<img width="829" height="334" alt="StaffInfoSection" src="https://github.com/user-attachments/assets/68c5a345-c702-4326-8be8-1f98573a0f26" />
<img width="577" height="126" alt="AddToCollectionContent" src="https://github.com/user-attachments/assets/478fd146-8fe5-46d5-af61-1bd7c96c81fb" />
<img width="594" height="233" alt="MinimizedInfoRow" src="https://github.com/user-attachments/assets/4a28eed6-4c4e-4f60-be12-18121c3e1418" />
<img width="503" height="553" alt="MainMovieOrSeriesDetails" src="https://github.com/user-attachments/assets/22dcd79a-8fe9-41f0-b875-f7d9c8a7dea9" />

Mohamed-Ashraf-2210 committed:
create entities for series details screen

Mohamed-Ashraf-2210 committed:
refactor app bar component

skaik-mo committed:
Add click handlers to AppBar back button and end icon

skaik-mo committed:
Merge branch 'fix/design-system-component' into components/fix-appbar

skaik-mo committed:
Merge pull request #171 from Rome-Squad/components/fix-appbar

Add click handlers to AppBar back button and end icon

skaik-mo committed:
Refactor: Update AppBar composable parameters and image display

- Renamed `showLogo` parameter to `showImage`.
- Added `image` and `imageSize` parameters to customize the displayed image.
- Updated default values for visibility parameters (`showBackButton`, `showImage`, `showTitle`, `showCaption`, `showEndIcon`) to `false`.
- Added horizontal padding to the AppBar's Row.
- Applied `CircleShape` clipping to the displayed image.

skaik-mo committed:
add `imageSourceToPainter` composable function

This function converts an image source (String URL or Int drawable resource) into a Jetpack Compose Painter. It uses `rememberAsyncImagePainter` for URLs and `painterResource` for drawable IDs.

Mohamed-Ashraf-2210 committed:
add image mapper

skaik-mo committed:
fix: Add social media click handlers to MainDetails

This commit introduces click handling for social media icons (YouTube, Facebook, Instagram) within the `MainDetails` composable.

- Added `onYoutubeClick`, `onFacebookClick`, and `onInstagramClick` lambda parameters to `MainDetails`.
- Made `SocialMediaComponent` clickable and invoked the `onClick` lambda when clicked.
- Updated string resources for actor birthday and place of birth to dynamically include the respective values.

skaik-mo committed:
fix: GalleryItem, GallerySection, and IconTextBox components

This commit introduces several updates to the GalleryItem, GallerySection, and IconTextBox components.

- `imageSourceToPainter.kt`: Refactored to be an extension function for `Any`.
- `GalleryItem.kt`:
    - Changed the `image` parameter type from `Int?` to `Painter?`.
- `IconTextBox.kt`:
    - Added a `size` modifier to the `Icon` with a value of `16.dp`.
- `GallerySection.kt`:
    - Added an `onShowMoreClick` lambda parameter.
    - Made the "Show More" text clickable.
    - Used `imageSourceToPainter` for gallery items.
    - Adjusted layout weights and modifiers for gallery items.
- `PreviewDetails.kt`:
    - Added `onShowMoreClick` to `GallerySectionPreview`.
- `strings.xml`: Capitalized "Gallery" and "Show More".

Mohamed-Ashraf-2210 committed:
fix some component like (MainInfoCard, MainMovieOrSeriesDetails, MinimizedInfoRow)

skaik-mo committed:
Refactor: SectionTitle and MoviesListSection composable

This commit refactors the `MoviesListSection` composable to utilize the newly introduced `SectionTitle` composable for displaying its title and optional clickable end text.

- The `SectionTitle` composable now handles the layout and styling of the section title and its associated clickable text.
- The `onClickShowMore` parameter in `MoviesListSection` has been renamed to `onClickEndText` for clarity.
- The `clickableText` parameter in `SectionTitle` is now nullable to support cases where no clickable text is needed.

skaik-mo committed:
Merge branch 'fix/design-system-component' into components/fix-appbar

* fix/design-system-component:
  fix some component like (MainInfoCard, MainMovieOrSeriesDetails, MinimizedInfoRow)
  add image mapper

# Conflicts:
#	presentation/details/src/main/java/com/giraffe/details/utils/imageSourceToPainter.kt

HendSayed25 committed:
delete all hard coded and add string file for ar

HendSayed25 committed:
fix some issues in cards

HendSayed25 committed:
reformat files

HendSayed25 committed:
add needed function to show the date in specific format

HendSayed25 committed:
updated recording to updates in components

HendSayed25 committed:
correct name of string

HendSayed25 committed:
Merge remote-tracking branch 'origin/fix/design-system-component' into fix/design-system-component

# Conflicts:
#	presentation/details/src/main/java/com/giraffe/details/preview/PreviewDetails.kt

skaik-mo committed:
Merge branch 'fix/design-system-component' into components/fix-appbar

* fix/design-system-component:
  correct name of string
  updated recording to updates in components
  add needed function to show the date in specific format
  reformat files
  fix some issues in cards
  delete all hard coded and add string file for ar

- **ReviewCard:**
    - Updated to use `formatAsMonthDayYear` for date display.
    - Added image source handling using `imageSourceToPainter`.
    - Made reviewerImageSource, reviewerName, and reviewerUsername non-nullable.
- **MainMovieOrSeriesDetails:**
    - Updated poster image source handling.
- **StaffInfoSection:**
    - Changed "Show More" text to be fetched from string resources.
    - Updated divider thickness.
    - Localized title in previews.
- **ReadMoreText:**
    - Updated preview with sample text.
- **Utils:**
    - Added `formatAsMonthDayYear` extension function for LocalDate.
- **SeasonCard:**
    - Added `onClick` parameter.
    - Updated divider thickness.
    - Localized "Rating", "Episodes", and "Year" text.
    - Updated poster image source handling in previews.
- **StarCastSection:**
    - Changed "Show More" text to be fetched from string resources.
- **MainDetails:**
    - Corrected "Instagram" string resource usage.
- **MinimizedInfoRow:**
    - Updated poster image source handling.
- **PreviewDetails:**
    - Updated various component previews with image source handling, localized text, and date formatting.
- **Localization:**
    - Added Arabic (ar-rEG) string resources for details components.
    - Updated English string resources.

# Conflicts:
#	presentation/details/src/main/java/com/giraffe/details/components/MainDetails.kt
#	presentation/details/src/main/res/values/strings.xml

Mohamed-Ashraf-2210 committed:
refactor component add to collection

Mohamed-Ashraf-2210 committed:
Merge branch 'fix/design-system-component' into components/fix-appbar

# Conflicts:
#	presentation/details/src/main/java/com/giraffe/details/components/MainDetails.kt
#	presentation/details/src/main/res/values/strings.xml

Mohamed-Ashraf-2210 committed:
Merge remote-tracking branch 'origin/components/fix-appbar' into components/fix-appbar

# Conflicts:
#	presentation/details/src/main/java/com/giraffe/details/components/MainDetails.kt
#	presentation/details/src/main/res/values/strings.xml

Mohamed-Ashraf-2210 committed:
resolve conflicts

Mohamed-Ashraf-2210 committed:
Merge pull request #173 from Rome-Squad/components/fix-appbar

Refactor AppBar, Sections, and Gallery Components

AhmedMohamed056 commented on presentation/details/src/main/java/com/giraffe/details/components/GalleryItem.kt:
```diff
@@ -22,7 +23,11 @@ import com.giraffe.designsystem.theme.Theme
+    image: Painter?,
+    contentDescription: String?,
+    modifier: Modifier = Modifier
+) {
```
make modifier the first parameter

AhmedMohamed056 commented on presentation/details/src/main/java/com/giraffe/details/components/MainInfoCard.kt:
```diff
@@ -0,0 +1,162 @@
+    rating: Float,
+    releaseDate: String,
+    modifier: Modifier = Modifier,
+    duration: String? = null,
```
make modifier the first parameter

skaik-mo replied:
Take a look at the `Image` component in `androidx.compose.foundation`, it places the required parameters first, followed by the `modifier`. Following this order improves readability and aligns with Compose conventions.

<img width="418" height="260" alt="image" src="https://github.com/user-attachments/assets/f3564188-9ded-4d1d-985f-c4b040a96708" />

AhmedMohamed056 commented on presentation/details/src/main/java/com/giraffe/details/components/MainInfoCard.kt:
```diff
@@ -0,0 +1,162 @@
+        modifier = modifier
+            .clip(RoundedCornerShape(Theme.radius.lg))
+            .background(Theme.color.background.card)
+            .padding(16.dp),
```
you can delete clip and add shape to background .background(color= Theme.color.background.card, shape= RoundedCornerShape(Theme.radius.lg))

skaik-mo replied:
Separating `shape` and `background` is fine, what’s your reason for combining the `shape` directly with `background` instead of keeping them separate?

poula22 replied:
@AhmedMohamed056 no actually he did it well as shape of background will clip the background itself not the component, so they are totally different.

AhmedMohamed056 replied:
@poula22 The difference between clip and shape in background appears when the component is clickable, so if the component is not clickable, there is not reason to use clip when component not clickable

AhmedMohamed056 commented on presentation/details/src/main/java/com/giraffe/details/components/MainInfoCard.kt:
```diff
@@ -0,0 +1,162 @@
+
+        Column(
+            verticalArrangement = Arrangement.spacedBy(11.5.dp),
+            horizontalAlignment = Alignment.CenterHorizontally,
```
are you sure that's padding 11.5?

skaik-mo replied:
Yup, it's 11.5 as per the design

AhmedMohamed056 commented on presentation/details/src/main/java/com/giraffe/details/components/MainInfoCard.kt:
```diff
@@ -0,0 +1,162 @@
+                modifier = Modifier
+                    .size(40.dp)
+                    .clip(RoundedCornerShape(Theme.radius.md))
+                    .background(Theme.color.button.primary)
```
the same here you can replace clip with shape in background

skaik-mo replied:
Separating `shape` and `background` is fine

poula22 replied:
@AhmedMohamed056 clip is better here also

AhmedMohamed056 commented on presentation/details/src/main/java/com/giraffe/details/components/MainInfoCard.kt:
```diff
@@ -0,0 +1,162 @@
+                modifier = Modifier
+                    .size(40.dp)
+                    .clip(RoundedCornerShape(Theme.radius.md))
+                    .background(Theme.color.button.secondary)
```
and the same

skaik-mo replied:
Separating `shape` and `background` is fine

poula22 replied:
@AhmedMohamed056 clip is better here also

AhmedMohamed056 commented on presentation/details/src/main/java/com/giraffe/details/components/MainMovieOrSeriesDetails.kt:
```diff
@@ -0,0 +1,77 @@
+    duration: String?,
+    releaseDate: String,
+    modifier: Modifier = Modifier
+) {
```
insure that modifier is the first in all components

skaik-mo replied:
Also here, required parameters should come before the `modifier` for consistency with Compose standards

AhmedMohamed056 commented on presentation/details/src/main/java/com/giraffe/details/components/MinimizedInfoRow.kt:
```diff
@@ -1,76 +1,106 @@
+    poster: Painter,
+    title: String,
+    modifier: Modifier = Modifier,
+    onClickPlay: () -> Unit = {},
```
and here

skaik-mo replied:
Also here, required parameters should come before the `modifier` for consistency with Compose standards

AhmedMohamed056 commented on presentation/details/src/main/java/com/giraffe/details/components/RatingStars.kt:
```diff
@@ -45,26 +45,20 @@ fun RatingStars(
+            ),
+                tint = if (isSelected) Theme.color.additional.primary.yellow
+                else Theme.color.shade.tertiary,
 
```
use animatedColorAsState

skaik-mo committed:
add color animation to rating stars

poula22 commented on designsystem/src/main/java/com/giraffe/designsystem/composable/AppBar.kt:
```diff
@@ -29,13 +33,17 @@ fun AppBar(
     modifier: Modifier = Modifier,
     title: String = "",
     caption: String = "",
+    image: Painter = painterResource(Theme.icons.colored.logo),
+    imageSize: Dp = 32.dp,
     endIcon: Painter = painterResource(Theme.icons.outline.add),
     hasBackground: Boolean = true,
-    showBackButton: Boolean = true,
-    showLogo: Boolean = true,
-    showTitle: Boolean = true,
-    showCaption: Boolean = true,
-    showEndIcon: Boolean = true,
+    showBackButton: Boolean = false,
+    showImage: Boolean = false,
+    showTitle: Boolean = false,
+    showCaption: Boolean = false,
+    showEndIcon: Boolean = false,
+    onBackButtonClick: () -> Unit = {},
+    onEndIconClick: () -> Unit = {},
```
I think a lot of default parameters need to be required for example:
title, caption, image, endIcon, onBackButtonClick, onEndIconClick

But it sill ok to approve just if you want to modify them it will be better to what is required parameters and what is not.

skaik-mo replied:
We have about 6 styles of app bar

poula22 commented on designsystem/src/main/java/com/giraffe/designsystem/composable/AppBar.kt:
```diff
@@ -45,39 +53,53 @@ fun AppBar(
+            modifier = Modifier
+                .size(24.dp)
+                .clickable(
+                    enabled = true,
```
useless

poula22 commented on designsystem/src/main/java/com/giraffe/designsystem/composable/MoviesListSection.kt:
```diff
@@ -26,36 +22,21 @@ fun MoviesListSection(
     paddingHorizontal: Int = 16,
-    onClickShowMore: () -> Unit = {},
+    onClickEndText: () -> Unit = {},
     onClickPoster: () -> Unit = {}
```
code style:
- required parameters
- required functions
- Modifier should be first optional parameter
- another optional parameter
- optional functions
- optional compose functions
- content compose function should be last optional or required parameter depending on your function.

poula22 commented on designsystem/src/main/java/com/giraffe/designsystem/composable/SectionTitle.kt:
```diff
@@ -2,22 +2,24 @@ package com.giraffe.designsystem.composable
+    clickableText: String? = null,
     modifier: Modifier = Modifier,
-    onClickableText: () -> Unit
+    onClickableText: () -> Unit = {}
```
please apply code style here also

poula22 commented on domain/series/src/main/java/com/giraffe/series/entity/Review.kt:
```diff
@@ -0,0 +1,13 @@
+    val userName: String,
+    val review: String,
+    val rating: Float,
+    val releaseYear: LocalDate?
```
creationDate is better here, should it be nullable?

poula22 commented on domain/series/src/main/java/com/giraffe/series/entity/SeriesDetails.kt:
```diff
@@ -0,0 +1,20 @@
+
+import kotlinx.datetime.LocalDate
+
+data class SeriesDetails(
+    val id: Int,
+    val posterUrl: String,
+    val name: String,
+    val genreIDs: List<SeriesGenre> = emptyList(),
+    val rating: Float,
+    val releaseYear: LocalDate?,
+    val overview: String,
+    val seasons: List<Season> = emptyList(),
+    val cast: List<Cast> = emptyList(),
+    val topOfReviews: List<Review> = emptyList(),
+    val director: String,
+    val characters: String,
+    val screenplay: String,
+    val story: String
+)
```
let's discuss this on slack

poula22 reviewed (dismissed):
Please apply code style and don't merge until we agree on entities

skaik-mo committed:
Refactor: Standardize modifier parameter order

This commit reorders the parameters in the `MoviesListSection`, `SectionTitle`, and `AppBar` composables to place the `modifier` parameter as the first optional parameter, adhering to common Kotlin coding conventions.

Specifically:
- In `MoviesListSection.kt`, the `modifier` parameter is moved before `endText`.
- In `SectionTitle.kt`, the `modifier` parameter is moved before `clickableText`.
- In `AppBar.kt`, the `enabled = true` parameter, which is the default, is removed from the `clickable` modifier on the back button icon for conciseness.

skaik-mo committed:
Replaced `viewModel` and `single` with the newer `viewModelOf` and `singleOf`

skaik-mo committed:
Refactor: implement animated for MainDetails

- Added `MainDetailsHeader` composable to display a compact header when scrolling.
- Integrated `SharedTransitionLayout` and `AnimatedContent` to animate the transition between the full details view and the header.
- The animation is triggered based on the scroll position.
- Refactored `MainDetails` to support shared element transitions for the actor image and name.
- Adjusted `AppBar` padding to accommodate the new header.
- Updated previews to showcase the animated header functionality.

skaik-mo committed:
Refactor: Make AppBar parameters nullable and remove redundant visibility flags

The `title`, `caption`, `image`, and `endIcon` parameters in the `AppBar` composable are now nullable. This change simplifies the API by removing the need for separate boolean flags (`showTitle`, `showCaption`, `showImage`, `showEndIcon`) to control their visibility.

mohamedebrahem13 committed:
Merge branch 'develop' into fix/design-system-component
